core/rio/hdt/src/main/java/org/eclipse/rdf4j/rio/hdt/HDTDictionarySectionPFC.java

```diff
@@ -181,4 +252,76 @@ protected void parse(InputStream is) throws IOException {
 		}
 		return arr;
 	}
+
+	/**
+	 * Encode a single block, removing the input strings to conserve memory
+	 * 
+	 * @param iter
+	 * @return encoded block
+	 */
+	private byte[] encodeBlock(Iterator<String> iter) {
+		byte[][] tmp = new byte[stringsBlock][];
+		int i = 0;
+
+		byte[] base = iter.next().getBytes(StandardCharsets.UTF_8);
+		iter.remove();
+		// copy base string into first entry, with a trailing NULL
+		tmp[i] = new byte[base.length + 1];
+		System.arraycopy(base, 0, tmp[i], 0, base.length);
+
+		byte[] prev = base;
+
+		// encode a block, with a maximum of strings per block (minus the first base string)
+		for (i = 1; iter.hasNext() && i < stringsBlock; i++) {
+			byte[] str = iter.next().getBytes(StandardCharsets.UTF_8);
+			iter.remove();
```

Comment:
Do we have to remove from the iterator? Just wondering if it can be skipped for performance.

Reply:
Well, I'm trying to somewhat reduce the memory consumption, so I'd rather keep it in here.

(I'll need to rewrite it anyway, since the actual goal is to compress very large datasets...)

Reply:
If your underlying data structure is an array list I think this causes the array list to be copied and the previous one to be garbage collected. 

Would be nice to actually benchmark this to see if it has any effect though. You are introducing a side effect that should be made obvious to the next developer that starts to wonder where all their data went.

Reply:
I see your point.

IIRC removing from an arraylist involves a memmove, not a copy, but I'm not entirely sure.
It would not require more memory, but involves (much) more memory shifting, so performance won't be great on larger sets :-/

Reply:
Removed the iter.remove, and using clear() in other parts of the code